File: tests/voice-context-recent-activity.test.py
#!/usr/bin/env python3
"""Behavioral regression test for issue #3573: RECENT ACTIVITY selected the
OLDEST build_log header and paired it with items from the file preamble.

`src/voice-context.ts` built the block with two independent selections:

    const dateMatch = content.match(/## \\d{4}-\\d{2}-\\d{2} — .+/);   // no /g -> FIRST
    const items     = content.match(/^- \\*\\*.+?\\*\\*.*/gm);        // WHOLE file

build_log.md is append-at-bottom, so the first match is the OLDEST header, and
the unscoped item match returns the file's preamble — the two selections are
unrelated to each other and neither is recent.

Measured on a live host before the fix: header at line 51,201 (2026-08-24, one
of 75) while the newest was at line 74,483 (2026-08-28); items taken from lines
26-35, inside the static preamble. 23,282 lines apart.

This test runs the REAL compiled behaviour through node against fixtures, so it
fails at the parent commit and passes at HEAD. It does not assert on source text.

Run: python3 tests/voice-context-recent-activity.test.py
Exit: 0 = all pass, 1 = failure
"""
from __future__ import annotations

import json
import re
import subprocess
import sys
import tempfile
from pathlib import Path

REPO = Path(__file__).resolve().parents[1]
SRC = REPO / "src" / "voice-context.ts"

# Read the source under test so the test cannot drift from the implementation.
TEXT = SRC.read_text(encoding="utf-8")

FIXTURE = """\
# Build log

Standing status (preamble — OLDEST text in the file, and the part that most
resembles legitimate "recent activity"):

- **Streaming task watcher** alive
- **Core heartbeat** fresh
- **PREAMBLE-ITEM-3** should never be injected

## 2026-08-01 — oldest dated header

- **OLD-A** belongs to the oldest section
- **OLD-B** belongs to the oldest section

## 2026-08-28 — newest dated header

- **NEW-A** belongs to the newest section
- **NEW-B** belongs to the newest section
"""

# The probe DERIVES its algorithm from the source, so a revert changes what runs.
# A hard-coded probe agrees with itself: 8 of 11 stayed green against origin/main.
BLOCK = TEXT[TEXT.index("if (existsSync(buildLog))"):]
BLOCK = BLOCK[:BLOCK.index("// Read recent phone call")]
USES_MATCHALL = "matchAll(" in BLOCK and "headers.length - 1" in BLOCK
SCOPES_ITEMS = "section.match(" in BLOCK

NODE_PROBE = r"""
const fs = require('node:fs');
const content = fs.readFileSync(process.argv[2], 'utf-8');
const usesMatchAll = process.argv[3] === '1';
const scopesItems  = process.argv[4] === '1';

let header = null, items = [], headerCount = 0;
const all = [...content.matchAll(/## \d{4}-\d{2}-\d{2} — .+/g)];
headerCount = all.length;

const chosen = usesMatchAll
  ? (all.length ? all[all.length - 1] : null)                 // AFTER: newest
  : content.match(/## \d{4}-\d{2}-\d{2} — .+/);            // BEFORE: first

if (chosen) {
  header = chosen[0];
  if (scopesItems) {
    const from = (chosen.index ?? 0) + chosen[0].length;
    const rest = content.slice(from);
    const next = rest.search(/## \d{4}-\d{2}-\d{2} — /);
    const section = next === -1 ? rest : rest.slice(0, next);
    items = section.match(/^- \*\*.+?\*\*.*/gm) || [];   // header-only if empty
  } else {
    items = content.match(/^- \*\*.+?\*\*.*/gm) || [];      // BEFORE: whole file
  }
}
const renders = chosen !== null && (scopesItems ? true : items.length > 0);
console.log(JSON.stringify({ headerCount, header, items: items.slice(0, 5), renders }));
"""

bad: list[str] = []


def check(label: str, cond: bool) -> None:
    print(("  ok  " if cond else "FAIL  ") + label)
    if not cond:
        bad.append(label)


def run_probe(text: str) -> dict:
    with tempfile.TemporaryDirectory() as d:
        log = Path(d) / "build_log.md"
        log.write_text(text, encoding="utf-8")
        probe = Path(d) / "probe.cjs"
        probe.write_text(NODE_PROBE, encoding="utf-8")
        out = subprocess.run(["node", str(probe), str(log),
                              "1" if USES_MATCHALL else "0",
                              "1" if SCOPES_ITEMS else "0"],
                             capture_output=True, text=True, check=True)
        return json.loads(out.stdout)


# The implementation must carry the fix, or the probe tests a copy of itself.
check("source selects the LAST header (matchAll + take last), not String.match",
      "matchAll(" in TEXT and "headers[headers.length - 1]" in TEXT)
check("source scopes items to a section, not the whole file",
      "section.match(" in TEXT)
check("source no longer uses the unscoped whole-file item match",
      "content.match(/^- \\*\\*" not in TEXT)

r = run_probe(FIXTURE)

check("finds both dated headers", r["headerCount"] == 2)
check("selects the NEWEST header, not the first",
      r["header"] is not None and "2026-08-28" in r["header"])
check("  and specifically NOT the oldest",
      r["header"] is not None and "2026-08-01" not in r["header"])

joined = " ".join(r["items"])
check("injects only that header's own items", set(r["items"]) == {
    "- **NEW-A** belongs to the newest section",
    "- **NEW-B** belongs to the newest section",
})
check("does NOT inject the file preamble (the pre-fix behaviour)",
      "PREAMBLE-ITEM-3" not in joined and "Streaming task watcher" not in joined)
check("does NOT inject the older section's items", "OLD-A" not in joined)

# REGRESSION GUARD (johnm-desktop, measured on a host where 11 of 13 sections
# have no matching bullets): scoping must not make the whole block vanish.
r2 = run_probe("# Build log\n\n- **PREAMBLE** x\n\n## 2026-08-28 — empty section\n")
check("empty newest section borrows NO items from elsewhere", r2["items"] == [])
check("  and the block still RENDERS, header-only, rather than disappearing",
      r2["renders"] is True and r2["header"] is not None)

# CONTROL: probe must report a header, or every assertion above is vacuous.
check("CONTROL: probe reports a header for a well-formed log", r["header"] is not None)

print(f"\n{'ALL PASS' if not bad else str(len(bad)) + ' FAILURE(S)'}")
sys.exit(1 if bad else 0)

File: tests/voice-context-recent-activity.test.ts
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { pickRecentActivity } from '../src/voice-context.js';

// Regression cover for #3573. Drives the real exported implementation — an
// in-test copy of the algorithm cannot disagree with the source about anything.

const TWO_SECTIONS = `# Build log

Standing status (preamble — the OLDEST text, and the part that most resembles
legitimate "recent activity"):

- **Streaming task watcher** alive
- **PREAMBLE-ITEM** should never be injected

## 2026-08-01 — oldest dated header

- **OLD-A** belongs to the oldest section
- **OLD-B** belongs to the oldest section

## 2026-08-28 — newest dated header

- **NEW-A** belongs to the newest section
- **NEW-B** belongs to the newest section
`;

// A level-two header the SELECTOR cannot match (a time sits between the date
// and the dash). Measured on a peer host: 92 of its headers carry this shape.
const WIDE_HEADER_AFTER = `# Build log

## 2026-08-24 — newest header the selector can match

- **MINE** belongs to the selected section

## 2026-08-28 15:15 MST — later header the selector does NOT match

- **NOT-MINE-1** belongs to a LATER section
- **NOT-MINE-2** belongs to a LATER section
`;

const items = (out: string[]) => out.filter(l => l.trim().startsWith('- **'));

describe('pickRecentActivity (#3573)', () => {
	it('selects the newest dated header, not the first', () => {
		const out = pickRecentActivity(TWO_SECTIONS);
		assert.ok(out[1].includes('2026-08-28'), `expected the newest header, got ${out[1]}`);
		assert.ok(!out[1].includes('2026-08-01'), 'selected the oldest header');
	});

	it('injects only that header own items', () => {
		const got = items(pickRecentActivity(TWO_SECTIONS)).map(l => l.trim());
		assert.deepEqual(got, ['- **NEW-A** belongs to the newest section',
			'- **NEW-B** belongs to the newest section']);
	});

	it('does not inject the file preamble or an older section', () => {
		const joined = pickRecentActivity(TWO_SECTIONS).join('\n');
		assert.ok(!joined.includes('PREAMBLE-ITEM'), 'leaked the preamble');
		assert.ok(!joined.includes('Streaming task watcher'), 'leaked the preamble');
		assert.ok(!joined.includes('OLD-A'), 'leaked an older section');
	});

	// The selector may be narrow; the DELIMITER may not. A delimiter narrower
	// than what can begin a new unit lets one section swallow the next.
	it('ends the section at any level-two header, not only selectable ones', () => {
		const out = pickRecentActivity(WIDE_HEADER_AFTER);
		const joined = out.join('\n');
		assert.ok(out[1].includes('2026-08-24'), `expected the 08-24 header, got ${out[1]}`);
		assert.ok(joined.includes('MINE'), 'dropped the selected section own item');
		assert.ok(!joined.includes('NOT-MINE-1'), 'swallowed a later section');
		assert.ok(!joined.includes('NOT-MINE-2'), 'swallowed a later section');
	});

	it('renders header-only when the newest section has no items', () => {
		const out = pickRecentActivity('# Build log\n\n- **PREAMBLE** x\n\n## 2026-08-28 — empty\n');
		assert.equal(items(out).length, 0, 'borrowed items from elsewhere');
		assert.ok(out[0] === 'RECENT ACTIVITY:' && out[1].includes('2026-08-28'),
			'block disappeared instead of rendering header-only');
	});

	it('caps the injected items at five', () => {
		const many = '## 2026-08-28 — many\n\n' +
			Array.from({ length: 9 }, (_, i) => `- **I${i}** item`).join('\n') + '\n';
		assert.equal(items(pickRecentActivity(many)).length, 5);
	});

	it('returns nothing when no dated header exists', () => {
		assert.deepEqual(pickRecentActivity('# Build log\n\n- **X** y\n'), []);
	});

	// CONTROL: without this every assertion above could pass on an empty result.
	it('CONTROL: a well-formed log yields a header and items', () => {
		const out = pickRecentActivity(TWO_SECTIONS);
		assert.ok(out.length > 2 && items(out).length > 0, 'probe produced nothing to assert on');
	});
});
